File: app/src/main/java/com/krystianwsul/checkme/gui/tasks/FriendPickerFragment.java
package com.krystianwsul.checkme.gui.tasks;


import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.gui.AbstractDialogFragment;

import junit.framework.Assert;

import java.util.List;

public class FriendPickerFragment extends AbstractDialogFragment {
    private ProgressBar mFriendPickerProgress;
    private RecyclerView mFriendPickerRecycler;

    @Nullable
    private List<FriendData> mFriendDatas;

    private Listener mListener;

    @NonNull
    public static FriendPickerFragment newInstance() {
        return new FriendPickerFragment();
    }

    public FriendPickerFragment() {

    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(getActivity())
                .title(R.string.friend_dialog_title)
                .customView(R.layout.fragment_friend_picker, false)
                .negativeText(android.R.string.cancel)
                .onNegative((dialog, which) -> dialog.cancel());

        MaterialDialog materialDialog = builder.build();

        LinearLayout linearLayout = (LinearLayout) materialDialog.getCustomView();
        Assert.assertTrue(linearLayout != null);

        mFriendPickerProgress = (ProgressBar) linearLayout.findViewById(R.id.friend_picker_progress);
        Assert.assertTrue(mFriendPickerProgress != null);

        mFriendPickerRecycler = (RecyclerView) linearLayout.findViewById(R.id.friend_picker_recycler);
        Assert.assertTrue(mFriendPickerRecycler != null);

        mFriendPickerRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));

        return materialDialog;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mFriendPickerProgress.setVisibility(View.GONE);
        mFriendPickerRecycler.setVisibility(View.VISIBLE);

        if (mFriendDatas != null)
            initialize();
    }

    public void initialize(@NonNull List<FriendData> friendDatas, @NonNull Listener listener) {
        mFriendDatas = friendDatas;
        mListener = listener;

        if (getActivity() != null)
            initialize();
    }

    private void initialize() {
        Assert.assertTrue(getActivity() != null);
        Assert.assertTrue(mFriendDatas != null);
        Assert.assertTrue(mListener != null);

        mFriendPickerRecycler.setAdapter(new FriendListAdapter());
    }

    public interface Listener {
        void onFriendSelected(@NonNull String friendId);
    }

    class FriendListAdapter extends RecyclerView.Adapter<FriendListAdapter.FriendHolder> {
        @Override
        public int getItemCount() {
            Assert.assertTrue(mFriendDatas != null);

            return mFriendDatas.size();
        }

        @Override
        public FriendListAdapter.FriendHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
            View friendRow = layoutInflater.inflate(R.layout.row_friend, parent, false);

            TextView friendName = (TextView) friendRow.findViewById(R.id.friendName);
            Assert.assertTrue(friendName != null);

            TextView friendEmail = (TextView) friendRow.findViewById(R.id.friendEmail);
            Assert.assertTrue(friendEmail != null);

            return new FriendListAdapter.FriendHolder(friendRow, friendName, friendEmail);
        }

        @Override
        public void onBindViewHolder(final FriendListAdapter.FriendHolder friendHolder, int position) {
            Assert.assertTrue(mFriendDatas != null);

            FriendData friendData = mFriendDatas.get(position);
            Assert.assertTrue(friendData != null);

            friendHolder.mFriendName.setText(friendData.mName);
            friendHolder.mFriendEmail.setText(friendData.mEmail);

            friendHolder.mFriendRow.setOnClickListener(v -> friendHolder.onRowClick());
        }

        class FriendHolder extends RecyclerView.ViewHolder {
            final View mFriendRow;
            final TextView mFriendName;
            final TextView mFriendEmail;

            FriendHolder(@NonNull View friendRow, @NonNull TextView friendName, @NonNull TextView friendEmail) {
                super(friendRow);

                mFriendRow = friendRow;
                mFriendName = friendName;
                mFriendEmail = friendEmail;
            }

            void onRowClick() {
                Assert.assertTrue(mFriendDatas != null);

                FriendData friendData = mFriendDatas.get(getAdapterPosition());
                Assert.assertTrue(friendData != null);

                dismiss();

                mListener.onFriendSelected(friendData.mId);
            }
        }
    }

    public static class FriendData {
        @NonNull
        final String mId;

        @NonNull
        final String mName;

        @NonNull
        final String mEmail;

        public FriendData(@NonNull String id, @NonNull String name, @NonNull String email) {
            Assert.assertTrue(!TextUtils.isEmpty(id));
            Assert.assertTrue(!TextUtils.isEmpty(name));
            Assert.assertTrue(!TextUtils.isEmpty(email));

            mId = id;
            mName = name;
            mEmail = email;
        }
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/friends/FriendPickerFragment.kt
package com.krystianwsul.checkme.gui.friends


import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import android.text.TextUtils
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.CustomItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.FragmentFriendPickerBinding
import com.krystianwsul.checkme.databinding.RowFriendBinding
import com.krystianwsul.checkme.gui.base.AbstractDialogFragment
import com.krystianwsul.checkme.gui.friends.findfriend.FindFriendActivity
import com.krystianwsul.checkme.gui.utils.ResettableProperty
import com.krystianwsul.checkme.utils.animateVisibility
import com.krystianwsul.checkme.utils.loadPhoto
import com.krystianwsul.common.utils.UserKey


class FriendPickerFragment : AbstractDialogFragment() {

    companion object {

        fun newInstance() = FriendPickerFragment()
    }

    private var data: Data? = null

    private lateinit var listener: (UserKey) -> Unit

    private val bindingProperty = ResettableProperty<FragmentFriendPickerBinding>()
    private var binding by bindingProperty

    @SuppressLint("InflateParams")
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        binding = FragmentFriendPickerBinding.inflate(layoutInflater)

        binding.friendPickerRecycler.layoutManager = LinearLayoutManager(activity)

        return MaterialAlertDialogBuilder(requireContext())
                .setTitle(R.string.friend_dialog_title)
                .setView(binding.root)
                .setPositiveButton(R.string.addFriend) { _, _ -> }
                .setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }
                .create()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        if (data != null) initialize()
    }

    override fun onStart() {
        super.onStart()

        (requireDialog() as AlertDialog).getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            startActivity(FindFriendActivity.newIntent(requireContext()))
        }
    }

    fun initialize(data: Data, listener: (UserKey) -> Unit) {
        this.data = data
        this.listener = listener

        if (bindingProperty.isSet) initialize()
    }

    private fun initialize() {
        check(activity != null)
        checkNotNull(data)

        animateVisibility(binding.friendPickerRecycler, binding.friendPickerProgress, data!!.immediate)

        binding.friendPickerRecycler.apply {
            adapter = FriendListAdapter()
            itemAnimator = CustomItemAnimator()
        }
    }

    override fun onDestroyView() {
        bindingProperty.reset()

        super.onDestroyView()
    }

    private inner class FriendListAdapter : RecyclerView.Adapter<FriendListAdapter.FriendHolder>() {

        override fun getItemCount() = data!!.friendDatas.size

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = FriendHolder(RowFriendBinding.inflate(layoutInflater, parent, false))

        override fun onBindViewHolder(friendHolder: FriendHolder, position: Int) {
            val friendData = data!!.friendDatas[position]

            friendHolder.binding.apply {
                friendImage.loadPhoto(friendData.photoUrl)
                friendName.text = friendData.name
                friendEmail.text = friendData.email

                root.setOnClickListener { friendHolder.onRowClick() }
            }
        }

        private inner class FriendHolder(val binding: RowFriendBinding) : RecyclerView.ViewHolder(binding.root) {

            fun onRowClick() {
                val friendData = data!!.friendDatas[adapterPosition]

                dismiss()

                listener(friendData.id)
            }
        }
    }

    data class Data(val immediate: Boolean, val friendDatas: List<FriendData>)

    class FriendData(
            val id: UserKey,
            val name: String,
            val email: String,
            val photoUrl: String?
    ) {

        init {
            check(!TextUtils.isEmpty(name))
            check(!TextUtils.isEmpty(email))
        }
    }
}
